File: kits/lambda-lifeline/src/iac/terraform.mjs
// Tokenize enough HCL to locate direct runtime attributes of top-level Lambda
// resource blocks. Comments, quoted text and heredocs never become configuration.
export function patchTerraform(text, from, to) {
  const tokens = [];
  let cursor = 0;
  while (cursor < text.length) {
    const rest = text.slice(cursor);
    const skipped = rest.match(/^(?:\s+|#[^\r\n]*|\/\/[^\r\n]*)/);
    if (skipped) { cursor += skipped[0].length; continue; }
    if (rest.startsWith('/*')) {
      const end = text.indexOf('*/', cursor + 2);
      if (end < 0) throw new Error('Unterminated Terraform comment; no changes written.');
      cursor = end + 2;
      continue;
    }
    const heredoc = rest.match(/^<<-?([A-Za-z_][\w-]*)[^\S\r\n]*\r?\n/);
    if (heredoc) {
      const bodyStart = cursor + heredoc[0].length;
      const end = new RegExp(`^[ \\t]*${heredoc[1]}[ \\t]*\\r?$`, 'm').exec(text.slice(bodyStart));
      if (!end) throw new Error('Unterminated Terraform heredoc; no changes written.');
      cursor = bodyStart + end.index + end[0].length;
      tokens.push({ value: '<heredoc>' });
      continue;
    }
    const quoted = rest.match(/^"(?:\\[\s\S]|[^"\\])*"/);
    if (rest[0] === '"' && !quoted) throw new Error('Unterminated Terraform string; no changes written.');
    const raw = quoted?.[0] || rest.match(/^[A-Za-z_][\w-]*/)?.[0] || rest[0];
    tokens.push({ value: quoted ? raw.slice(1, -1) : raw, quoted: Boolean(quoted), start: cursor, end: cursor + raw.length });
    cursor += raw.length;
  }
  const blocks = [];
  const matches = [];
  for (let i = 0; i < tokens.length; i++) {
    const token = tokens[i];
    if (!token.quoted && token.value === '{') {
      blocks.push(blocks.length === 0 && tokens[i - 3]?.value === 'resource'
        && tokens[i - 2]?.quoted && tokens[i - 2]?.value === 'aws_lambda_function'
        && Boolean(tokens[i - 1]?.quoted));
    } else if (!token.quoted && token.value === '}') {
      if (!blocks.length) throw new Error('Unbalanced Terraform braces; no changes written.');
      blocks.pop();
    } else if (blocks.length === 1 && blocks[0] && !token.quoted && token.value === 'runtime'
      && tokens[i + 1]?.value === '=' && tokens[i + 2]?.quoted
      && from.includes(tokens[i + 2].value) && tokens[i + 2].value !== to) {
      matches.push(tokens[i + 2]);
    }
  }
  if (blocks.length) throw new Error('Unbalanced Terraform braces; no changes written.');
  let changed = text;
  for (const token of [...matches].reverse()) {
    changed = changed.slice(0, token.start) + JSON.stringify(to) + changed.slice(token.end);
  }
  return { changed, hits: matches.map(token => token.value) };
}

File: kits/python-pivot/src/python_pivot/terraform.py
"""Byte-local edits of direct runtime attributes in Lambda HCL resource blocks."""

import re


def patch_terraform(text, source_runtimes, target_runtime):
    tokens = []
    cursor = 0
    while cursor < len(text):
        rest = text[cursor:]
        skipped = re.match(r"(?:\s+|#[^\r\n]*|//[^\r\n]*)", rest)
        if skipped:
            cursor += len(skipped[0])
            continue
        if rest.startswith("/*"):
            end = text.find("*/", cursor + 2)
            if end < 0:
                raise ValueError("Unterminated Terraform comment; no changes written")
            cursor = end + 2
            continue
        heredoc = re.match(r"<<-?([A-Za-z_][\w-]*)[^\S\r\n]*\r?\n", rest)
        if heredoc:
            body_start = cursor + len(heredoc[0])
            end = re.search(
                r"^[ \t]*" + re.escape(heredoc[1]) + r"[ \t]*\r?$", text[body_start:], re.MULTILINE
            )
            if not end:
                raise ValueError("Unterminated Terraform heredoc; no changes written")
            cursor = body_start + end.end()
            tokens.append(("<heredoc>", False, cursor, cursor))
            continue
        quoted = re.match(r'"(?:\\[\s\S]|[^"\\])*"', rest)
        if rest[0] == '"' and not quoted:
            raise ValueError("Unterminated Terraform string; no changes written")
        identifier = re.match(r"[A-Za-z_][\w-]*", rest)
        raw = quoted[0] if quoted else identifier[0] if identifier else rest[0]
        tokens.append((raw[1:-1] if quoted else raw, bool(quoted), cursor, cursor + len(raw)))
        cursor += len(raw)
    blocks = []
    matches = []
    for i, (value, quoted, _, _) in enumerate(tokens):
        if not quoted and value == "{":
            is_lambda = (
                not blocks
                and i >= 3
                and tokens[i - 3][0] == "resource"
                and tokens[i - 2][:2] == ("aws_lambda_function", True)
                and tokens[i - 1][1]
            )
            blocks.append(is_lambda)
        elif not quoted and value == "}":
            if not blocks:
                raise ValueError("Unbalanced Terraform braces; no changes written")
            blocks.pop()
        elif blocks == [True] and not quoted and value == "runtime" and i + 2 < len(tokens):
            candidate = tokens[i + 2]
            if (
                tokens[i + 1][0] == "="
                and candidate[1]
                and candidate[0] in source_runtimes
                and candidate[0] != target_runtime
            ):
                matches.append(candidate)
    if blocks:
        raise ValueError("Unbalanced Terraform braces; no changes written")
    changed = text
    for _, _, start, end in reversed(matches):
        changed = changed[:start] + '"' + target_runtime + '"' + changed[end:]
    return changed, len(matches)
